Feature/all items

serpient opened the pull request:
# Creates All Items Page
All Items Page uses a new fetch utility to make an API call for items data. The data is then set using ReactHooks, rendered as ItemCard components, then placed as child components in the ItemsPagePresenter component.

- Create fetch utility that will make a get request given a endpoint. Can be used standalone, or in a useEffect() call in the following below example format. As a hint, when using useEffect(), be sure to set the second argument to determine when it should execute. Not setting it correctly will cause the function trigger numerous times. Read more here -> https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects
```
const [items, setItems] = useState([]);

    useEffect(() => {
        const fetchItems = async () => {
            let result = await Fetch("/api/items");
           setItems(result)
        }
        fetchItems();
    }, []);
```
- Add tests for ItemsPageContainer using axiosMock and testing-library/react. Currently only asserting that the correct number of data objects will be rendered.
- Create starter shared scss variables file
- Create UI Page modular component
- Each component is split into Container (logic) and Presenter (presentational).

# Things To Do
- Add edge case tests for Error / Loading states in All Items Page & ItemCard

# Live Shot
<img width="699" alt="Screen Shot 2019-07-17 at 11 02 56 PM" src="https://user-images.githubusercontent.com/29721784/61433029-11bcac00-a8e7-11e9-8d0a-b22e1bc26a56.png">

serpient committed:
Add Router setup

To prove everything is setup for local development, quick and simple ItemsPage and ItemPage components were created. ItemsPage uses useEffect and axios to ensure our API call uses proxy correctly when the backend server is running locally. Now that it's proven to work, we can scrap and start again using TDD.

- All Items Page has "/" and "/items" route
- SIngle Item Page has the following route "/item/:id"
- Add a localhost:4000 proxy which will route API calls only during development. We will still need a proper setup for the base endpoint for deployment.
- Add Axios dependency to make API calls

serpient committed:
Create All Items page

All Items Page uses a new fetch utility to make an API call for items data. The data is then set using ReactHooks, rendered as ItemCard components, then placed as child components in the ItemsPagePresenter component.

- Create fetch utility that will make a get request given a endpoint. Can be used standalone, or in a useEffect() call in the following below example format. As a hint, when using useEffect(), be sure to set the second argument to determine when it should execute. Not setting it correctly will cause the function trigger numerous times. Read more here -> https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects
```
const [items, setItems] = useState([]);

    useEffect(() => {
        const fetchItems = async () => {
            let result = await Fetch("/api/items");
           setItems(result)
        }
        fetchItems();
    }, []);
```
- Add tests for ItemsPageContainer using axiosMock and testing-library/react. Currently only asserting that the correct number of data objects will be rendered.
- Create starter shared scss variables file
- Create UI Page modular component
- Each component is split into Container (logic) and Presenter (presentational).

serpient committed:
Merge branch 'feature/router' into feature/all_items

tpetersen0308 commented on src/App.js:
```diff
@@ -13,7 +12,6 @@ function App() {
         <Route path="/" exact component={ItemsPage} />
         <Route path="/items" exact component={ItemsPage} />
-        <Route path="/item/:id" exact component={ItemPage} />
       </Router>
```
This route should be `/items/:id`

serpient committed:
Add error handling to ItemsPage container

zacholauson commented on .vscode/launch.json:
```diff
@@ -0,0 +1,16 @@
+{
```
I don't know how vscode functions, but do we need both `launch.json` files? Also, JSON doesn't have a valid comments format so they should probably be removed from there.

jayden replied:
As an aside, should `.vscode` be ignored completely?

zacholauson commented on src/__mocks__/axios.js:
```diff
@@ -0,0 +1,9 @@
+const axios = jest.genMockFromModule('axios');
+
+const mockAxios = {
```
https://8thlight.com/blog/eric-smith/2011/10/27/thats-not-yours.html
https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own

serpient replied:
I refactored to create a APIHandler that can take in a Fetcher and MockFetcher class

zacholauson commented on src/components/items/Presenter.test.js:
```diff
@@ -0,0 +1,9 @@
+import ReactDOM from 'react-dom';
+import ItemsPagePresenter from './Presenter';
+
+it('renders without crashing', () => {
```
Should we be asserting something about how `Presenter` renders its children?

jayden commented on src/components/items/Container.js:
```diff
@@ -0,0 +1,37 @@
+    }, []);
+
+    const createItemComponents = (data) => {
+        return data.map((object, index) => {
```
It's an [anti-pattern](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318) to use the index when creating component lists. I'd suggest using some other unique identifier like the item's id.

zacholauson commented on src/utilities/fetch.test.js:
```diff
@@ -0,0 +1,45 @@
+    expect.assertions(3);
+    expect(axiosMock.get).toHaveBeenCalledTimes(1)
+    expect(axiosMock.get).toHaveBeenCalledWith(url)
+    expect(fetchResult).toEqual(itemData.data)
```
This doesn't seem like an overly useful test since it's really just testing that the mock returns what you told it to return.

zacholauson commented on src/components/item_card/Presenter.js:
```diff
@@ -0,0 +1,24 @@
+
+const ItemCard = ({ item }) => {
+    let { title, price, image, end_date } = item;
+    const dateFormatter = (endDate) => {
```
Are there any tests for this?

serpient replied:
We extracted this out and added tests in the `show_item` PR

serpient committed:
Refactor per feedback

- Ignore vscode files
- Add render tests for Presenter, asserting that children elements are rendered
- Remove numbered index in render maps

serpient committed:
Create APIHandler with fetcher and mockFetcher

- APIHandler has 2 duck types: Fetcher which uses axios to correspond with the API, and MockFetcher that takes in data during its construction and returns it.
- Each logical component should take in ApiHandler as a prop (dependency injection) to make it easier to test each component.

serpient committed:
Rename components to match refactor in /show_item

`/show_item` branch used this branch (`/all_items`) as a base.  We are renaming here in hopes of reducing the amount of merge conflicts that will occur when `/show_item` needs to rebase this branch.
